File: src/test/kotlin/com/bemsa/scraper/controllers/advice/RestExceptionHandlerTest.java
package com.bemsa.scraper.controllers.advice;

import com.bemsa.scraper.exceptions.DataNotFoundException;
import com.bemsa.scraper.exceptions.GitScraperException;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.http.ResponseEntity;

class RestExceptionHandlerTest {
    private final RestExceptionHandler handler = new RestExceptionHandler();

    @Test
    void testHandleDataNotFoundException() {
        //when
        var errorResponse = handler.handleDataNotFoundException(new DataNotFoundException(""));

        //then
        asserts(errorResponse);
    }

    @Test
    void testHandleGitScraperException() {
        //when
        var errorResponse = handler.handleGitScraperException(new GitScraperException(""));

        //then
        asserts(errorResponse);
    }

    @Test
    void testHandleCatchAllException() {
        //when
        var errorResponse = handler.handleCatchAllException(new RuntimeException(""));

        //then
        asserts(errorResponse);
    }

    private void asserts(ResponseEntity<ApiResponse> responseEntity) {
        Assertions.assertThat(responseEntity)
                .isNotNull()
                .isExactlyInstanceOf(ResponseEntity.class)
                .isNotEqualTo(ResponseEntity.ok().build());
    }
}

File: src/test/kotlin/com/bemsa/scraper/controllers/advice/RestExceptionHandlerTest.kt
package com.bemsa.scraper.controllers.advice

import com.bemsa.scraper.exceptions.DataNotFoundException
import com.bemsa.scraper.exceptions.GitScraperException
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.http.ResponseEntity

class RestExceptionHandlerTest {
    private val handler = RestExceptionHandler()

    @Test
    fun testHandleDataNotFoundException() {
        //when
        val errorResponse = handler.handleDataNotFoundException(DataNotFoundException("Data not found."))

        //then
        asserts(errorResponse, ResponseEntity.notFound().build())
    }

    @Test
    fun testHandleGitScraperException() {
        //when
        val errorResponse = handler.handleGitScraperException(GitScraperException("Bad request was sent."))

        //then
        asserts(errorResponse, ResponseEntity.badRequest().build())
    }

    @Test
    fun testHandleCatchAllException() {
        //when
        val errorResponse = handler.handleCatchAllException(RuntimeException("An error has occurred!"))

        //then
        asserts(errorResponse, ResponseEntity.internalServerError().build())
    }

    private fun asserts(responseEntity: ResponseEntity<ApiResponse> , error: ResponseEntity<String>) {
        assertThat(responseEntity)
                .isNotNull
                .isExactlyInstanceOf(ResponseEntity::class.java)
                .isNotEqualTo(error)
    }
}
